[TourMarkerView] Added the ability to select between elapsed, moving and recorded time

FJBDev opened the pull request:
Per [user request](https://sourceforge.net/p/mytourbook/discussion/622811/thread/17da3fb19f/?page=1#32dc), adding the ability to select between elapsed, moving and recorded time.

![image](https://user-images.githubusercontent.com/25235665/140189816-9f7cb473-0283-454f-a244-27b0595acf45.png)

wolfgang-ch commented:
> 
> 
> Per [user request](https://sourceforge.net/p/mytourbook/discussion/622811/thread/17da3fb19f/?page=1#32dc), adding the ability to select between elapsed, moving and recorded time.
> 
> ![image](https://user-images.githubusercontent.com/25235665/140189816-9f7cb473-0283-454f-a244-27b0595acf45.png)

`Use ... time`  for what ?

I think the UI should be similar like

![image](https://user-images.githubusercontent.com/1283445/140264738-f8c67efc-58df-424b-bb6e-1ce34a6c6d5a.png)

with a label, e.g. `Compute laps time with` (when I've read the request correctly)

![image](https://user-images.githubusercontent.com/1283445/140265122-0a2da71c-b5d7-4227-b2d6-52aad02578a6.png)

and with a tooltip e.g. "Compute time between markers..." or we should label it differently, "Laps time" do not everybody know, it is mostly used by runners.

wolfgang-ch commented:
> 
> 
> Per [user request](https://sourceforge.net/p/mytourbook/discussion/622811/thread/17da3fb19f/?page=1#32dc), adding the ability to select between elapsed, moving and recorded time.
> 
> ![image](https://user-images.githubusercontent.com/25235665/140189816-9f7cb473-0283-454f-a244-27b0595acf45.png)

Remove "View" from the title "Tour Marker View Options"

FJBDev commented:
> I think the UI should be similar like
> 
> with a label, e.g. `Compute laps time with` (when I've read the request correctly)
> 
> and with a tooltip e.g. "Compute time between markers..." or we should label it differently, "Laps time" do not everybody know, it is mostly used by runners.

Fixed.
I renamed it to "Time used between markers" because it's used for:
- time
- time difference
- average speed
- average pace
![image](https://user-images.githubusercontent.com/25235665/140429552-4cf13bbf-8495-48bb-bbdb-f9a2194f8ea4.png)


I also added a tooltip for better explanation.

![image](https://user-images.githubusercontent.com/25235665/140429599-1cae3488-db0a-4e6d-b574-07ec46bb0239.png)


>Remove "View" from the title "Tour Marker View Options"
Fixed. See above

wolfgang-ch commented:
Almost perfect, the tooltip text "time/time ..." irritated me first until I understood it "time / time difference / ..."
For easier reading you could also use multiple lines like in the above text on this web page

```
The selected time will be used to compute the 

* Time
* Time difference
* Speed
* Pace 
 
between markers
```

FJBDev commented:
Fixed

![image](https://user-images.githubusercontent.com/25235665/140584580-75153b96-365c-490b-bd4b-214c7bd4b22b.png)

wolfgang-ch reviewed (commented):
Merged with minor issues

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourMarker/SlideoutTMVOptions.java:
```diff
@@ -0,0 +1,207 @@
+      }
+   }
+
+   /*
+    * UI controls
+    */
```
A button is not a UI control ?

wolfgang-ch replied:
For me, this not a comment for a method

![image](https://user-images.githubusercontent.com/1283445/142753372-1c42e39c-9d86-42eb-b110-bf67a7fc21ab.png)

FJBDev replied:
I was looking at the one at the top... yes, this one is from a copy and paste. I will remove it in my next PR

FJBDev replied:
Fixed

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourMarker/TourMarkerView.java:
```diff
@@ -98,34 +100,36 @@
-
    private PostSelectionProvider    _postSelectionProvider;
    private ISelectionListener       _postSelectionListener;
    private IPropertyChangeListener  _prefChangeListener;
    private IPropertyChangeListener  _prefChangeListener_Common;
    private ITourEventListener       _tourEventListener;
-   private IPartListener2           _partListener;
 
+   private IPartListener2           _partListener;
    private MenuManager              _viewerMenuManager;
-   private IContextMenuProvider     _tableViewerContextMenuProvider = new TableContextMenuProvider();
 
+   private IContextMenuProvider     _tableViewerContextMenuProvider = new TableContextMenuProvider();
    private ActionOpenMarkerDialog   _actionEditTourMarkers;
+
    private ActionDeleteMarkerDialog _actionDeleteTourMarkers;
 
-   private PixelConverter           _pc;
+   private ActionToolbarSlideout    _actionTmvOptions;
 
+   private PixelConverter           _pc;
    private TableViewer              _markerViewer;
-   private ColumnManager            _columnManager;
 
+   private ColumnManager            _columnManager;
    private boolean                  _isInUpdate;
-   private boolean                  _isMultipleTours;
 
+   private boolean                  _isMultipleTours;
    private ColumnDefinition         _colDefName;
-   private ColumnDefinition         _colDefVisibility;
 
+   private ColumnDefinition         _colDefVisibility;
    private final NumberFormat       _nf3                            = NumberFormat.getNumberInstance();
```
Reverted in my current [PR](https://github.com/wolfgang-ch/mytourbook/pull/422)

FJBDev replied:
is there an explanation why Eclipse sometimes reformats code but sometime doesn't ?

wolfgang-ch replied:
I've not yet searched for this issue but this happens since many years and may depend on the structure of the fields and it's embedded {...} blocks

FJBDev commented on bundles/net.tourbook/src/net/tourbook/ui/views/tourMarker/TourMarkerView.java:
```diff
@@ -137,6 +141,7 @@
     */
    private PageBook  _pageBook;
    private Composite _pageNoData;
+
    private Composite _viewerContainer;
 
    private Font      _boldFont;
```
Reverted in my current [PR](https://github.com/wolfgang-ch/mytourbook/pull/422)

FJBDev commented:
@wolfgang-ch 
It seems to me that there is still an issue with how my commits are merged in main (and I think i mentioned that in the past several times)
If you look at the [files modified](https://github.com/wolfgang-ch/mytourbook/commit/bb76816104d0aab429468a7d8faac5bd6ea99bf6), it shows the version modification that belongs to [your commit here](https://github.com/wolfgang-ch/mytourbook/commit/9b0ec2a760291f34d6e005f53fa4119270980ee4) and that was already committed to main 9h before you merged my PR.
I think I mentioned that before, to me, it creates a lot of confusion because if I (or anyone else...) look in my commits in the future, it will look like I have modified those files but i actually did not. Also, that is not clear because my commit describes my specific changes (here my TourMarkerView changes) but those changes are lost in other changes. And why do those files show changed when they were already updated in your commit ?

I know you don't want to but i think one solution would be to do PR against main. Maybe there are other solutions that I don't know of.

![image](https://user-images.githubusercontent.com/25235665/142745049-c7479de1-6a52-4f2a-8753-b794ad775267.png)

wolfgang-ch commented:
> @wolfgang-ch It seems to me that there is still an issue with how my commits are merged in main (and I think i mentioned that in the past several times) If you look at the [files modified](https://github.com/wolfgang-ch/mytourbook/commit/bb76816104d0aab429468a7d8faac5bd6ea99bf6), it shows the version modification that belongs to [your commit here](https://github.com/wolfgang-ch/mytourbook/commit/9b0ec2a760291f34d6e005f53fa4119270980ee4) and that was already committed to main 9h before you merged my PR (why do those files show changed when they were already updated in your commit ?). I think I mentioned that before, to me, it creates a lot of confusion because if I (or anyone else...) look in my commits in the future, it will look like I have modified those files but i actually did not. Also, that is not clear because my commit describes my specific changes (here my TourMarkerView changes) but those changes are lost in other changes.
> 
> I know you don't want to but i think one solution would be to do PR against main. Maybe there are other solutions that I don't know of.

Before I do a merge, I check the network graph from where the PR was lastly updated

![image](https://user-images.githubusercontent.com/1283445/142752995-d0b9e0bc-338a-444b-bb30-809811519802.png)

Here is my gitlog, I can send you the exact date/time values when you need it

![image](https://user-images.githubusercontent.com/1283445/142753037-bee9916f-740c-4711-b1bc-8e80be247da6.png)

FJBDev commented:
> Here is my gitlog, I can send you the exact date/time values when you need it
> 
I already can access the gitlog [here](https://github.com/wolfgang-ch/mytourbook/commits/main)

wolfgang-ch commented:
> @wolfgang-ch It seems to me that there is still an issue with how my commits are merged in main (and I think i mentioned that in the past several times) If you look at the [files modified](https://github.com/wolfgang-ch/mytourbook/commit/bb76816104d0aab429468a7d8faac5bd6ea99bf6), it shows the version modification that belongs to [your commit here](https://github.com/wolfgang-ch/mytourbook/commit/9b0ec2a760291f34d6e005f53fa4119270980ee4) and that was already committed to main 9h before you merged my PR (why do those files show changed when they were already updated in your commit ?). I think I mentioned that before, to me, it creates a lot of confusion because if I (or anyone else...) look in my commits in the future, it will look like I have modified those files but i actually did not. Also, that is not clear because my commit describes my specific changes (here my TourMarkerView changes) but those changes are lost in other changes.
> 
> I know you don't want to but i think one solution would be to do PR against main. Maybe there are other solutions that I don't know of.
> 


**Before merge**

After I had reset FJBDev to main, I pressed several times the F5 browser refresh button (on the right side of the screen) but the PR did not recognize the updated FJBDev branch in my repo. I did this also with the previous PR (and created screenshots) and this PR had updated to my current FJBDev. So it seems that GitHub is not always working as it should.


![mt before merge](https://user-images.githubusercontent.com/1283445/143995452-decdb7df-43f2-4280-b2fe-c387859ba8c8.png)

**After merge**

![mt after merge](https://user-images.githubusercontent.com/1283445/143995456-a8cc7df5-b0d2-4a20-ae8f-34d963693b66.png)


>I know you don't want to but i think one solution would be to do PR against main. 

When after a review (which can span several hours or days), I would not accept a PR which is merged into main, I had to revert main which could cause problems when other people had synced their git repo. And I need to merge a PR to review it with Eclipse, the Github review feature is horror for me when there are more than a few line changes.

>Maybe there are other solutions that I don't know of.

There could be (git is so much powerfull) but I don't know another solutions

FJBDev commented:
> When after a review (which can span several hours or days), I would not accept a PR which is merged into main, I had to revert main which could cause problems when other people had synced their git repo. And I need to merge a PR to review it with Eclipse, the Github review feature is horror for me when there are more than a few line changes.

But why do you merge into main before reviewing ? The standard practice I think is to merge when the review is finished, not the other way around.
If you look at most of the PRs on Github (or Gitlab..etc), they are mostly done against main/master to avoid all those "hoops" ([example](https://github.com/microsoft/vscode-dev-containers/pull/1179)).
It also has the advantage that what is merged is the final "good" code in one single commit instead of having multiple commits like those ones where the first one is not perfect, then there is a 2nd one, then a 3rd one...etc until it is good, which to me pollutes a little the commit history.

![image](https://user-images.githubusercontent.com/25235665/144091395-08569a9c-90de-409c-82bf-e5e8f5769fb0.png)


Maybe you can create the source branch locally (in this case `fb-markerviewslideout`) to do the review ?

wolfgang-ch commented:
I will give you access to the FJBDev branch in my repo, you should have received an invitation.

I don't yet know which access levels you would have when you accect the invitation but you must have access only to the FJBDev branch and not to others.

FJBDev commented:
> I will give you access to the FJBDev branch in my repo, you should have received an invitation.
> 
> I don't yet know which access levels you would have when you accect the invitation but you must have access only to the FJBDev branch and not to others.

I just clicked on the invitation link and I see that:
![image](https://user-images.githubusercontent.com/25235665/144320127-afaba15b-348a-4e9e-973f-8fb86158ba1d.png)

It doesn't say that i only have push to FJBDev though is that ok ?

Also, does that mean you want me to put my new code on FJBDev and then create a Pull Request from FJBDev to main ?

FJBDev commented:
@wolfgang-ch 
I think there is something wrong because I was able to create a new branch in your repository. I should only have access to FJBDev branch

wolfgang-ch commented:
> @wolfgang-ch I think there is something wrong because I was able to create a new branch in your repository. I should only have access to FJBDev branch I have removed my created branch that was called `fb-weathertests`
> 

Because this is the first time to give access to the MT repo to someone else, I was not sure which access level I can set. After you have accepted the invitation, now I can see, that I cannot set any access levels for you.

![image](https://user-images.githubusercontent.com/1283445/144372349-431664b3-43f1-46fe-a28a-dec33901d9e9.png)


> Maybe you should remove my access to your repo, that sounds too dangerous to be able to create branches or modify your `main` branch😨

Yesterday I found out how to protect the `main` branch :-)

![image](https://user-images.githubusercontent.com/1283445/144372787-fb76059c-99d5-48ca-a69f-a56ac02bc68d.png)

wolfgang-ch commented:
> > I will give you access to the FJBDev branch in my repo, you should have received an invitation.
> > I don't yet know which access levels you would have when you accect the invitation but you must have access only to the FJBDev branch and not to others.
> 
> I just clicked on the invitation link and I see that: ![image](https://user-images.githubusercontent.com/25235665/144320127-afaba15b-348a-4e9e-973f-8fb86158ba1d.png)
> 
> It doesn't say that i only have push to FJBDev though is that ok ?

I cannot set access levels to branches but branches can be protected with PR's, see above screenshot
 
> Also, does that mean you want me to put my new code on FJBDev and then create a Pull Request from FJBDev to main ?

Yes, now that your code is in my repo, I can use Eclipse (without copying your repo) to review your PR before merging it.

wolfgang-ch commented:
When I try with the current branch rule to commit into the main branch, I've no access, I would also need to create a PR.

After reading some GitHub docus, I found this:

![image](https://user-images.githubusercontent.com/1283445/144396375-38c7ae3b-0458-479f-bf59-138c382a22d0.png)

So I will try to convert the MT repo into an organisation.

wolfgang-ch commented:
> > @wolfgang-ch I think there is something wrong because I was able to create a new branch in your repository. I should only have access to FJBDev branch I have removed my created branch that was called `fb-weathertests`
> 
> Because this is the first time to give access to the MT repo to someone else, I was not sure which access level I can set. After you have accepted the invitation, now I can see, that I cannot set any access levels for you.
> 
> ![image](https://user-images.githubusercontent.com/1283445/144372349-431664b3-43f1-46fe-a28a-dec33901d9e9.png)
> 

After moving MT into an organisation, there are more options for the user access levels

![image](https://user-images.githubusercontent.com/1283445/144441910-06ab82c4-193d-4b7e-9d8d-47e082512a6a.png)

FJBDev commented:
> So I will try to convert the MT repo into an organisation.

I have accepted the invitation
![image](https://user-images.githubusercontent.com/25235665/144452912-b83d2428-1b32-4802-9b33-665360954c74.png)

FJBDev commented:
> I cannot set access levels to branches but branches can be protected with PR's, see above screenshot

Ok, as long as I only modify FJBDev, we should be good!


> Yes, now that your code is in my repo, I can use Eclipse (without copying your repo) to review your PR before merging it.

Perfect, then my [PR](https://github.com/mytourbook/mytourbook/pull/426) is ready to be reviewed and merged
